addons/sale/models/sale.py

```diff
@@ -259,10 +259,15 @@ def onchange_partner_id_warning(self):
     @api.model
     def create(self, vals):
         if vals.get('name', _('New')) == _('New'):
-            if 'company_id' in vals:
-                vals['name'] = self.env['ir.sequence'].with_context(force_company=vals['company_id']).next_by_code('sale.order') or _('New')
-            else:
-                vals['name'] = self.env['ir.sequence'].next_by_code('sale.order') or _('New')
+            ctx = self.env.context.copy()
+            if vals.get('company_id'):
+                ctx['force_company'] = vals['company_id']
+            if vals.get('date_order'):
+                ctx['ir_sequence_date'] = self._fields['date_order'].convert_to_cache(vals['date_order'], False, validate=False)[:10]
+            sequence_obj = self.env['ir.sequence']
+            if ctx:
+                sequence_obj = sequence_obj.with_context(ctx)
+            vals['name'] = sequence_obj.next_by_code('sale.order') or _('New')
```

Comment:
Actually, we this work ? 
```
if vals.get('name', _('New')) == _('New'):
	Sequence = self.env['ir.sequence'].with_context(ir_sequence_date=vals.get('date_order'))
    if 'company_id' in vals:
        Sequence = Sequence.with_context(force_company=vals['company_id'])

    vals['name'] = Sequence.next_by_code('sale.order') or _('New')
```

Reply:
That's the problem: vals.get('date_order') might contain:

- A full datetime compliant string. OK.
- A non full datetime string, like in certain demo data that doesn't contain seconds. KO. 
- A datetime object, like in certain tests. KO.

I have discovered this in a refinement process when runbot has failed because of these cases (you can see the forced pushes now in GitHub). That's why I normalize the value through this convoluted expression. Maybe you know a better way...